src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2055,458 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Metadata syncing is only allowed for hash, "
+							   "reference and local tables: %c", partitionMethod)));
+	}
+
+	List *distShardTupleList = LookupDistShardTuples(relationId);
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		if (!(shardMinValue == NULL && shardMaxValue == NULL))
+		{
+			char *relationName = get_rel_name(relationId);
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Shards of reference or local table \"%s\" should "
+								   "have NULL shard ranges", relationName)));
+		}
+		else if (list_length(distShardTupleList) != 0)
+		{
+			char *relationName = get_rel_name(relationId);
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("relation \"%s\" has already at least one shard, "
+								   "adding more is not allowed", relationName)));
+		}
+	}
+	else if (partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		if (shardMinValue == NULL || shardMaxValue == NULL)
+		{
+			char *relationName = get_rel_name(relationId);
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Shards of has distributed table  \"%s\" "
+								   "cannot have NULL shard ranges", relationName)));
+		}
+
+		char *shardMinValueString = text_to_cstring(shardMinValue);
+		char *shardMaxValueString = text_to_cstring(shardMaxValue);
+
+		/* pg_strtoint32 does the syntax and out of bound checks for us */
+		int32 shardMinValueInt = pg_strtoint32(shardMinValueString);
+		int32 shardMaxValueInt = pg_strtoint32(shardMaxValueString);
+
+		if (shardMinValueInt >= shardMaxValueInt)
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("shardMinValue=%d is greater or equal to "
+								   "shardMaxValue=%d for table \"%s\", which is "
+								   "not allowed", shardMinValueInt,
+								   shardMaxValueInt, get_rel_name(relationId))));
+		}
+
+		/*
+		 * We are only dealing with hash distributed tables, that's why we
+		 * can hard code data type and typemod.
+		 */
+		const int intervalTypeId = INT4OID;
+		const int intervalTypeMod = -1;
+
+		Relation distShardRelation = table_open(DistShardRelationId(), AccessShareLock);
+		TupleDesc distShardTupleDesc = RelationGetDescr(distShardRelation);
+
+		FmgrInfo *shardIntervalCompareFunction =
+			GetFunctionInfo(intervalTypeId, BTREE_AM_OID, BTORDER_PROC);
+
+		HeapTuple shardTuple = NULL;
+		foreach_ptr(shardTuple, distShardTupleList)
+		{
+			ShardInterval *shardInterval =
+				TupleToShardInterval(shardTuple, distShardTupleDesc,
+									 intervalTypeId, intervalTypeMod);
+
+			Datum firstMin = Int32GetDatum(shardMinValueInt);
+			Datum firstMax = Int32GetDatum(shardMaxValueInt);
+			Datum secondMin = shardInterval->minValue;
+			Datum secondMax = shardInterval->maxValue;
+			Oid collationId = InvalidOid;
+
+			if (ShardIntervalsOverlapWithParams(firstMin, firstMax, secondMin, secondMax,
```

Comment:
Are we certain here that `secondMin` and `secondMax` are non NULL?

Reply:
yes, we have `if (shardMinValue == NULL || shardMaxValue == NULL)` check above

Reply:
that's only applicable to `firstMin` and `firstMax` right? `secondMin` and `secondMax` come from the rows from `pg_dist_shard`. Although I guess, because of that check, we never insert a NULL value in that table for hash distributed tables, so we should be good here.

Reply:
>  Although I guess, because of that check, we never insert a NULL value in that table for hash distributed tables, so we should be good here.

Oh, yes that was my assumption. Might be good to add an explicit check or assert